File: android/src/main/kotlin/twilio/flutter/twilio_programmable_chat/methods/ChannelsMethods.kt
package twilio.flutter.twilio_programmable_chat.methods

import com.twilio.chat.CallbackListener
import com.twilio.chat.Channel
import com.twilio.chat.ChannelDescriptor
import com.twilio.chat.ErrorInfo
import com.twilio.chat.Paginator
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import twilio.flutter.twilio_programmable_chat.Mapper
import twilio.flutter.twilio_programmable_chat.PaginatorManager
import twilio.flutter.twilio_programmable_chat.TwilioProgrammableChatPlugin

object ChannelsMethods {
    fun createChannel(call: MethodCall, result: MethodChannel.Result) {
        val friendlyName = call.argument<String>("friendlyName")
                ?: return result.error("ERROR", "Missing 'friendlyName'", null)

        val channelTypeValue = call.argument<String>("channelType")
                ?: return result.error("ERROR", "Missing 'channelType'", null)

        val channelType = when (channelTypeValue) {
            "PRIVATE" -> Channel.ChannelType.PRIVATE
            "PUBLIC" -> Channel.ChannelType.PUBLIC
            else -> null
        } ?: return result.error("ERROR", "Wrong value for 'channelType'", null)

        try {
            TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.createChannel(friendlyName, channelType, object : CallbackListener<Channel>() {
                override fun onSuccess(newChannel: Channel) {
                    TwilioProgrammableChatPlugin.debug("ChannelsMethods.createChannel => onSuccess")
                    result.success(Mapper.channelToMap(newChannel))
                }

                override fun onError(errorInfo: ErrorInfo) {
                    TwilioProgrammableChatPlugin.debug("ChannelsMethods.createChannel => onError: $errorInfo")
                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                }
            })
        } catch (err: IllegalArgumentException) {
            return result.error("IllegalArgumentException", err.message, null)
        }
    }

    fun getChannel(call: MethodCall, result: MethodChannel.Result) {
        val channelSidOrUniqueName = call.argument<String>("channelSidOrUniqueName")
                ?: return result.error("ERROR", "Missing 'channelSidOrUniqueName'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSidOrUniqueName, object : CallbackListener<Channel>() {
            override fun onSuccess(newChannel: Channel) {
                TwilioProgrammableChatPlugin.debug("ChannelsMethods.getChannel => onSuccess")
                result.success(Mapper.channelToMap(newChannel))
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("ChannelsMethods.getChannel => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getPublicChannelsList(call: MethodCall, result: MethodChannel.Result) {
        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getPublicChannelsList(object : CallbackListener<Paginator<ChannelDescriptor>>() {
            override fun onSuccess(paginator: Paginator<ChannelDescriptor>) {
                TwilioProgrammableChatPlugin.debug("ChannelsMethods.getPublicChannelsList => onSuccess")
                val pageId = PaginatorManager.setPaginator(paginator)
                result.success(Mapper.paginatorToMap(pageId, paginator, "channelDescriptor"))
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("ChannelsMethods.getPublicChannelsList => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getUserChannelsList(call: MethodCall, result: MethodChannel.Result) {
        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getUserChannelsList(object : CallbackListener<Paginator<ChannelDescriptor>>() {
            override fun onSuccess(paginator: Paginator<ChannelDescriptor>) {
                TwilioProgrammableChatPlugin.debug("ChannelsMethods.getUserChannelsList => onSuccess")
                val pageId = PaginatorManager.setPaginator(paginator)
                result.success(Mapper.paginatorToMap(pageId, paginator, "channelDescriptor"))
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("ChannelsMethods.getUserChannelsList => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getMembersByIdentity(call: MethodCall, result: MethodChannel.Result) {
        val identity = call.argument<String>("identity")
                ?: return result.error("ERROR", "Missing 'identity'", null)

        val memberList = TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getMembersByIdentity(identity)
        val membersListMap = memberList?.map { Mapper.memberToMap(it) }
        result.success(membersListMap)
    }
}

File: ios/Classes/Methods/ChannelsMethods.swift
import Flutter
import Foundation
import TwilioChatClient

public class ChannelsMethods {
    public static func createChannel(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let friendlyName = arguments["friendlyName"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'friendlyName' parameter", details: nil))
        }

        guard let channelTypeString = arguments["channelType"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'channelType' parameter", details: nil))
        }

        guard let channelType = Mapper.stringToChannelType(channelTypeString) else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Could no parse 'channelType' parameter", details: nil))
        }

        let channelOptions: [String: Any?] = [
            TCHChannelOptionFriendlyName: friendlyName,
            TCHChannelOptionType: channelType.rawValue
        ]

        let flutterResult = result
        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.createChannel(options: channelOptions, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                SwiftTwilioProgrammableChatPlugin.debug("ChannelsMethods.createChannel => onSuccess")
                flutterResult(Mapper.channelToDict(channel))
            } else {
                SwiftTwilioProgrammableChatPlugin.debug("ChannelsMethods.createChannel => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error creating channel with friendlyName '\(friendlyName)': \(String(describing: result.error))", details: nil))
            }
        })
    }

    public static func getChannel(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let channelSidOrUniqueName = arguments["channelSidOrUniqueName"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'channelSidOrUniqueName' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSidOrUniqueName, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                SwiftTwilioProgrammableChatPlugin.debug("ChannelsMethods.getChannel => onSuccess")
                flutterResult(Mapper.channelToDict(channel))
            } else {
                SwiftTwilioProgrammableChatPlugin.debug("ChannelsMethods.getChannel => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel with sid or uniqueName '\(channelSidOrUniqueName)'", details: nil))
            }
        })
    }

    public static func getPublicChannelsList(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let flutterResult = result
        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.publicChannelDescriptors(completion: { (result: TCHResult, paginator: TCHChannelDescriptorPaginator?) in
            if result.isSuccessful(), let paginator = paginator {
                SwiftTwilioProgrammableChatPlugin.debug("ChannelsMethods.getPublicChannelsList => onSuccess")
                let pageId = PaginatorManager.setPaginator(paginator)
                flutterResult(Mapper.channelDescriptorsToDict(pageId, paginator))
            } else {
                SwiftTwilioProgrammableChatPlugin.debug("ChannelsMethods.getPublicChannelsList => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving public channels list", details: nil))
            }
        })
    }

    public static func getUserChannelsList(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let flutterResult = result
        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.userChannelDescriptors(completion: { (result: TCHResult, paginator: TCHChannelDescriptorPaginator?) in
            if result.isSuccessful(), let paginator = paginator {
                SwiftTwilioProgrammableChatPlugin.debug("ChannelsMethods.getUserChannelsList => onSuccess")
                let pageId = PaginatorManager.setPaginator(paginator)
                flutterResult(Mapper.channelDescriptorsToDict(pageId, paginator))
            } else {
                SwiftTwilioProgrammableChatPlugin.debug("ChannelsMethods.getUserChannelsList => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving user channels list", details: nil))
            }
        })
    }

    public static func getMembersByIdentity(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let identity = arguments["identity"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'identity' parameter", details: nil))
        }

        var membersList: [[String: Any?]] = []
        // TODO: Get all channels - Public and User and get members across them
        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.subscribedChannels().forEach({ (channel: TCHChannel) in
            if let member = channel.member(withIdentity: identity) {
                membersList.append(Mapper.memberToDict(member, channelSid: channel.sid))
            }
        })
        result(membersList)
    }
}
